File: 36.valid-sudoku.java
/*
 * @lc app=leetcode id=36 lang=java
 *
 * [36] Valid Sudoku
 */
import java.util.HashSet;
import java.util.Set;

// @lc code=start
class Solution {
    public boolean isValidSudoku(char[][] board) {
        // check horizontal
        boolean isValidX = checkHorizontal(board);
        // check vertical
        boolean isValidY = checkVertical(board);
        // check each 3 x 3 square
        int sqSize = 3;
        for (int row = 0; row < board.length; row += sqSize) {
            for (int col = 0; col < board[0].length; col += sqSize) {
                if (!checkSubBox(board, row, col, sqSize)) {
                    return false;
                }
            }
        }
        return isValidX && isValidY;
    }

    private boolean checkHorizontal(char[][] board) {
        for (int row = 0; row < board.length; row++) {
            // check each row for unique
            Set<Character> setChars = new HashSet<>();
            for (int col = 0; col < board[row].length; col++) {
                char currentChar = board[row][col];
                if (currentChar == '.') {
                    continue;
                }
                else if (setChars.contains(currentChar)) {
                    return false;
                }
                else {
                    setChars.add(currentChar);
                }
            }
        }
        return true;
    }

    private boolean checkVertical(char[][] board) {
        for (int col = 0; col < board.length; col++) {
            // check each col for unique
            Set<Character> setChars = new HashSet<>();
            for (int row = 0; row < board[col].length; row++) {
                char currentChar = board[row][col];
                if (currentChar == '.') {
                    continue;
                }
                else if (setChars.contains(currentChar)) {
                    return false;
                }
                else {
                    setChars.add(currentChar);
                }
            }
        }
        return true;
    }
    
    private boolean checkSubBox(char[][] board, int row, int col, int sqSize) {
        Set<Character> setChars = new HashSet<>();
        for (int i = row; i < row + sqSize; i++) {
            for (int j = col; j < col + sqSize; j++) {
                char currentChar = board[i][j];
                if (currentChar == '.') {
                    continue;
                }
                else if (setChars.contains(currentChar)) {
                    return false;
                }
                else {
                    setChars.add(currentChar);
                }
            }
        }
        return true;
    }
}
// @lc code=end

File: 36.valid-sudoku.js
/*
 * @lc app=leetcode id=36 lang=javascript
 *
 * [36] Valid Sudoku
 */

// @lc code=start
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function (board) {
  let len = board.length;
  // check horizontal
  for (let i = 0; i < len; i++) {
    let rowNums = [];
    for (let j = 0; j < len; j++) {
      if (board[i][j] != ".") {
        rowNums.push(board[i][j]);
      }
    }
    if (!isUniqueDigits(rowNums)) {
      return false;
    }
  }
  // check vertical
  for (let i = 0; i < len; i++) {
    let colNums = [];
    for (j = 0; j < len; j++) {
      if (board[j][i] != ".") {
        colNums.push(board[j][i]);
      }
    }
    if (!isUniqueDigits(colNums)) {
      return false;
    }
  }
  // check subsquare
  let row = 0;
  let col = 0;
  let count = 0;
  let squareNums = [];
  for (let i = row; i < row + 3; i++) {
    for (let j = col; i < col + 3; j++) {
      squareNums.push(board[row][col]);
    }
  }
  if (!isUniqueDigits(squareNums)) {
    return false;
  }
  count++;
  if (count < 3) {
    row = 0;
    col += 3;
  }
  if (count < 6) {
    row = 3;
    col += 3;
  }
};

const isUniqueDigits = (digits) => {
  let seen = {};

  for (let digit of digits) {
    if (seen.hasOwnProperty(digit)) {
      return false;
    } else {
      seen[digit] = 1;
    }
  }
  return true;
};
// @lc code=end
